### PR TITLE
LanceDB integration #7839

### LINKED ISSUE #7839: [Integration]: Integrate LanceDB (vector db) with MindsDB
### Is there an existing integration?

- [X] I have searched the existing integrations.

### Use Case

The goal of this issue is to build a new handler that will connect MindsDB to LanceDB vector database.

Add support for LanceDB vector database

### Motivation

We already support Chroma, lets add support for LanceDB

### Implementation

Resources 💡
🏁 To get started, please refer to the following resources:

📚 [Building Data Handler docs](https://docs.mindsdb.com/contribute/data-handlers): This documentation will guide you through the process of building a custom handler for MindsDB. It contains essential information on the handler structure, methods, and best practices.

📖 [LanceDB](https://lancedb.com/)

👉 [ChromaDB](https://github.com/mindsdb/mindsdb/tree/staging/mindsdb/integrations/handlers/chromadb_handler): This existing handler connects MindsDB with ChromaDB.

Next Steps 🕵️‍♂️ 🕵️‍♀️
You can familiarize yourself with LanceDB.
Create a new handler that connects MindsDB to LanceDB.
Implement the core features mentioned above.
Thoroughly test the integration to ensure proper functionality.
Submit a pull request with the new LanceDB Integration handler.

### Anything else?

📢 Additional rewards 🏆
Each App integration brings 2️⃣ 5️⃣ : points for SWAG and entry into the draw for a 👇 :

💻 [Razer Blade 16 Laptop](https://www.razer.com/gaming-laptops/Razer-Blade-16/RZ09-0483SEJ3-R3U1)
👕 🐻 [MindsDB Swag](https://store.covver.io/mindsdb/collections/swag-collection)
ℹ️ For more info check out https://mindsdb.com/hacktoberfest/ 👈

### PR DESCRIPTION
**Pull request by bispaul**
## Description

Handler for [LanceDB](https://lancedb.com/)

Fixes #7839

## Type of change

(Please delete options that are not relevant)

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] 📄 This change requires a documentation update

## Verification Process

To ensure the changes are working as expected:

 - [ ]   Test Location: Specify the URL or path for testing.
 - [x]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.

## Additional Media:

- [x] I have attached a brief loom video or screenshots showcasing the new functionality or change.

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [x] Necessary documentation updates are either made or tracked in issues.
- [x] Relevant unit and integration tests are updated or added.

## Testing LanceDB Handler

**1. . Testing LanceDB CREATE DATABASE**
```
CREATE DATABASE lancedb
WITH ENGINE = "lancedb",
PARAMETERS = {
    "persist_directory" : "~/lancedb"
};
```
Screenshot 2023-10-29 at 2 49 40 AM

Screenshot 2023-10-29 at 2 50 03 AM

**2. Testing CREATE TABLE**
```
CREATE TABLE lancedb.test_data18
(SELECT *
FROM myexample_db.lance_test_data2);
```
Screenshot 2023-10-29 at 2 50 44 AM

**3. Testing SELECT FROM**
```
select * from lancedb.test_data18;

select * from lancedb.test_data18
where id = 'id1';
```
Screenshot 2023-10-29 at 2 51 21 AM
Screenshot 2023-10-29 at 2 51 44 AM

**4. Testing INSERT**
```
INSERT INTO lancedb.test_data18 (id,content,metadata,embeddings)
VALUES ("id9", 'this is a test', '{"test":"test4"}', '[5,2,3,2,3,4,2,9]');

select * from lancedb.test_data18
where id ='id9';
```
Screenshot 2023-10-29 at 2 52 19 AM
Screenshot 2023-10-29 at 2 52 46 AM

**5: Testing Distance**

```
select * from lancedb.test_data18
where search_vector = '[5,2,3,2,3,4,2,9]'
;
```
Screenshot 2023-10-29 at 2 54 26 AM

**6. Testing DELETE**
```
delete from lancedb.test_data18 where id ='id9';

select * from lancedb.test_data18;
```

Screenshot 2023-10-29 at 2 55 27 AM
Screenshot 2023-10-29 at 2 56 05 AM

**7: Unit Test**
 
Screenshot 2023-10-29 at 2 46 42 AM

Drop a remark based on your observation.

-   [x] Works Great 💚 (This means that all the steps were executed successfuly and the expected outputs were returned.)
-   [ ] There's a Bug 🪲 [Issue Title](URL To the Issue you created) ( This means you encountered a Bug. Please open an issue with all the relevant details with the Bug Issue Template)

---

### CODE REVIEW

**Comment by bispaul** [comment]
have read the CLA Document and I hereby sign the CLA

File: mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        self.is_connected = False
+
+    def check_connection(self):
+        """Check the connection to the LanceDB database."""
+        response_code = StatusResponse(False)
+        need_to_close = self.is_connected is False
+
+        try:
+            # self._client.heartbeat()
+            response_code.success = True
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py:91]: This doesn't check the connection? Do we need the heartbeat uncommented?

File: mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+        Insert data into the LanceDB database.
+        """
+        try:
+            collection = self._client.open_table(table_name)
+            df = data[[TableField.ID.value, TableField.CONTENT.value, TableField.METADATA.value, TableField.EMBEDDINGS.value]]
+            pa_data = pa.Table.from_pandas(df, preserve_index=False)
+            vec_data = vec_to_table(df[TableField.EMBEDDINGS.value].values.tolist())
+            new_pa_data = pa_data.append_column("vector", vec_data["vector"])
+            collection.add(new_pa_data)
+        except pa.lib.ArrowNotImplementedError:
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py:240]: This looks confusing. What is happening here? Can you also add detailed comments to the insert method?

File: mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        if limit and offset:
+            sql = f"""select {col_str} from result {where_str} limit {limit} offset {offset}"""
+        elif limit and not offset:
+            sql = f"""select {col_str} from result {where_str} limit {limit}"""
+        elif offset and not limit:
+            sql = f"""select {col_str} from result {where_str} offset {offset}"""
+        else:
+            sql = f"""select {col_str} from result {where_str}"""
+
+        data_df = duckdb.query(sql).to_df()
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py:224]: Do we need duckdb as a dependency just to query the df?

**Review by ZoranPandovski — CHANGES_REQUESTED** [review]

File: mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py (hunk 2 above)
Comment (bispaul) [reply mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py:240]: Hi @ZoranPandovski , when a pandas dataframe is sent to a lancedb dataset which is empty (0 rows), the column type does not usually match and the exception pa.lib.ArrowNotImplementedError is thrown, in that case the dataset is recreated (only when the initial dataset is empty). Let me know if you have any other idea in mind.

File: mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py (hunk 1 above)
Comment (bispaul) [reply mindsdb/integrations/handlers/lancedb_handler/lancedb_handler.py:90]: Hi @ZoranPandovski there is no self._client.heartbeat() in lancedb. I fixed that piece and pushed.

**Comment by bispaul** [comment]
HI @ZoranPandovski , @duanzhihua any updates?

**Review by dusvyat — APPROVED** [review]
thanks for your patience, lgtm

**Pull request merged**